src/logger.coffee

```diff
@@ -0,0 +1,155 @@
+_ = require 'lodash'
+PUBNUB = require 'pubnub'
+Promise = require 'bluebird'
+es = require 'event-stream'
+Lock = require 'rwlock'
+{ checkTruthy } = require './lib/validation'
+
+LOG_PUBLISH_INTERVAL = 110
+
+# Pubnub's message size limit is 32KB (unclear on whether it's KB or actually KiB,
+# but we'll be conservative). So we limit a log message to 2 bytes less to account
+# for the [ and ] in the array.
+MAX_LOG_BYTE_SIZE = 30000
+MAX_MESSAGE_INDEX = 9
+
+module.exports = class Logger
+	constructor: ({ @eventTracker }) ->
+		@publishQueue = [[]]
+		@messageIndex = 0
+		@publishQueueRemainingBytes = MAX_LOG_BYTE_SIZE
+		@logsOverflow = false
+		_lock = new Lock()
+		@_writeLock = Promise.promisify(_lock.async.writeLock)
+		@attached = {}
+		@offlineMode = false
+		@publishEnabled = false
+
+	init: ({ pubnub, channel, offlineMode, enable }) =>
+		Promise.try =>
```

Comment:
This would be nicer with `Promise.method`

Reply:
Promise.method doesn't work well for any of these methods because they wouldn't allow me to use a fat arrow properly.
Here, `this` inside the init function refers to the logger instance, whereas if I did:
```
  init: Promise.method (...) =>
```
Then `this` inside the init function would refer to the Logger *class* because the Promise.method would run in the context of the class definition. From the coffee-script v1 docs:
>  [...] in the context of a class definition, this is the class object itself [...]

Reply:
I don't think that section of the coffee-script docs is relevant here, and I'm pretty sure you can use a thin arrow unless you use something like
```coffee
logger = new Logger()
{ init } = logger
init()
```
or
```coffee
logger = new Logger()
logger.init.call(null)
```
if you use it as
```coffee
logger = new Logger()
logger.init()
```
then it'll be fine

Reply:
It would break any time the function is called inside a Promise chain like `Promise.try(logger.init)`.
Example:
```
Promise = require 'bluebird'
class A
        constructor: ->
                console.log('created an A')
                @importantProperty = 'someValue'
        slimArrowFn: ->
                console.log(@importantProperty)
        fatArrowFn: =>
                Promise.try =>
                        console.log(@importantProperty)
        methodFn: Promise.method ->
                console.log(@importantProperty)
        methodFatArrowFn: Promise.method =>
                console.log(@importantProperty)

a = new A()
Promise.try(a.slimArrowFn)
.then(a.fatArrowFn)
.then(a.methodFn)
.then(a.methodFatArrowFn)
```
Output:
```
pablo@pablo-laptop:~/resin/resin-containers/src/resin-supervisor$ coffee test.coffee 
created an A
undefined
someValue
undefined
undefined
```

So only by using a fat arrow do we ensure that we can always access the instance's properties from within the method without having to manually bind the method by doing `Promise.try(a.slimArrowFn.bind(a))` or `Promise.try( -> a.slimArrowFn())`.
So I'd rather keep the way it is to avoid this accidentally happening at some point in the future.

Reply:
Is that something you actually make use of?

Reply:
Maybe not for this particular function, but in general we prefer to do `.then(fn)` rather than `.then( -> fn())`, so I'd rather as a rule ensure that my methods work when we do that (otherwise we'll be scratching our heads when something breaks because we did that simple change)

Reply:
And by extension, I'd rather have all of the instance methods bound to the instance with a fat arrow, unless it's explicitly a method that I want to be able to call on a different object at some point (which is rare)

Reply:
(All the instance methods that use any `@property`, that is)